Android overhaul (includes important bug fixes and new functionality)

Benjamin-Dobell opened the pull request:
This is a mammoth pull request, and I know that's a pain for maintainers. So I've tried to make this as easily digestible as possible. I've elaborated on the motivation for the changes below, and although I didn't write it that way, I've gone back and split this into smaller feature commits so they can each be evaluated some-what individually.

- c967bff484f3313420429e0e3c7ab76c3669aa5b - Migrate to FCM and refactor

This is the largest commit of the lot, but mostly does what it says. The motivation for upgrading from GCM (deprecated) to FCM is pretty obvious but some of the refactoring probably needs further explanation/justification.

In particular I've completely done away with the concept of `PushNotification` and split its functionality into two classes, `LocalNotification` and `RemoteNotification`, which are slightly more limited in scope.

This is a **breaking change** to the Android Java API.

Additionally `PushNotificationProps` has been changed to `NotificationProps` to highlight its dual role and had its behaviour adjusted to more accurately represent Google/Firebase's concept of a notification. In particular, Firebase `RemoteMessage` class does _not_ bundle end-user data arguments in with UI notification properties. Notification properties and "data" are distinct concepts.

`NotificationProps` contains a `getData()` method to return the _data only_, the notification properties can be accessed by their corresponding accessor methods. This flows over into JS where there is a property called `data`, which should take a JS object and all user/custom data should be part of that object, not as part of the top-level notification. This is also a **breaking change**.

These breaking changes weren't made arbitrarily or even to keep the project neat. They're actually _required_ to facilitate correct function. In particular we need to be able to reliably determine whether a notification is "data only" so that JS consumers know under which context a notification is being delivered to them. I believe this is something that was perhaps missed by the original author of this functionality, but **there are very distinct rules under which notifications are delivered** to running applications. I've actually explained these rules in a README update (later commit) but the [official Firebase reference](https://firebase.google.com/docs/cloud-messaging/android/receive) explains in detail.

The final **breaking change** is that `RemoteNotification` does _not_ automatically post a local notification when a push/remote notification is received (in the foreground). This is important as it gives JS consumers control over when they'd like to display notifications and also what information they include in said notifications. Firebase does not support several options (large icons, LED lighting) etc.

Android now supports `cancelAllLocalNotifications`.

- ae36e8a3b89ebddbf31743870beeab03230f420b - Namespace globally broadcast JS events

Fairly self explanatory, this is necessitated by the fact Android's JS events are broadcast "globally", so it's best to avoid any accidental interference by name-spacing.

- 521f0c7e3bd9d4003f9361b63267987764fd1e55 - Support FCM token invalidation

A simple mechanism to invalidate push tokens. This can be used for example when a user logs out to ensure the app stops receiving notifications.

- 98424119cfb51eacaf8ee1503dab9fbed3eb2ee2 - Fixed initial intent behavior for background notifications

I've covered this pretty thoroughly in the README and it relates to the "data-only" changes mentioned above (first commit) - or in this case, when the notification _isn't_ data-only. If the system tray receives our push notification because the app was backgrounded, at the moment there's no way to tell when the user launched the app by tapping that notification - this commit corrects that situation.

- faf442e27c5e27989e51fdd90a8fa510fc6db4d0 - Do not automatically dismiss all notifications

This is about giving control to JS consumers. At present any time the app is launched all notifications are dismissed. That's likely to be undesirable for many use cases, particularly when a user has received multiple push notifications pertaining to different components of the app. The support that was added (first commit) for `cancelAllLocalNotifications()` means users can still achieve this effect if it's desirable. This is a **breaking change** with respect to default expected behaviour.

- 812ced326fd8cece8850dee02ed879e0558297c8 - Android large icon support (URL or drawable name)

This is actually pretty dang nifty. It's now trivial to specify `largeIcon` appears in the notification (typically used in messaging apps as profile pictures). Fresco (React Native's image manager) is used to download the images if they're specified as remote URLs - it takes care of caching and what-not.

Additionally, `file://` URLs are supported, which is what React Native converts `require("./some_image.jpg")` syntax to in JS. So React Native bundled images can seamlessly be used in addition to remote images.

- de8c172949c76531dbb2c04f96fdfde6c2bc48ea - Minor code simplification/clean-up

Just a nit. Nothing particularly important or pressing.

- f084ebd5b3cfcb55b8422d5ab7aaec01a14d24b4 - Move getInitialNotification() NotificationsAndroid (match iOS)

Got rid of `PendingNotifications`, as it was an unnecessary namespace. This is a **breaking change**.

- 59c6237552ac4bd130d83fb9120d8576b3f86ccb - Android lights/LED support

Added support for controlling the LED notification light i.e. What Facebook and similar apps do. 

Firebase (system tray) remote notifications are actually incapable of achieving this, so it's one of many very good reasons to send "data-only" notifications and generate local notifications from the content.

- 7ff0123cd8c3c672aca7026d5e412ebb77c7d7b4 - Android README updates

- dc3e4626f00fb2bb0b987360b60f438f6b07f760 - Expose invalidateToken() to JS

Expose invalidateToken() implemented in 521f0c7e3bd9d4003f9361b63267987764fd1e55 to JS.

- 3c1f7a72c43e47621f6dadd4e34c6f02ee2f4e72 - Fixed native Android test compilation

Ensure the Android native tests compile correctly with the latest changes.

- 835cd866461d72ff4663410ad927100c96faac33 - Refactored ProxyService into LocalNotificationService

General clean-up and removal of dead/unnecessary code.

- ff3ae152b3f57be9fb2ba2f29572214ad642974b - Android background queue

Fixed (or rather added) proper background notification support. Android needs a background queue, for precisely the same reasons iOS needs one. Therefore, I've added `NotificationsAndroid.consumeBackgroundQueue()`.

Events won't be delivered at all unless this is called, which is a **breaking change**, but the same behaviour as iOS. I've updated the documentation accordingly. In doing so I've also addressed some misleading advice / sample code. In particular the (iOS) documentation showed adding and removing event listeners inside components. This is incorrect, "cold start" background notifications will not immediately be received by your listener if they're added inside a component, as the component hierarchy does not exist at all until your app is foregrounded.

- a207aa4235f6e6ed85463e81d06b889dda7ad50d - RN 0.47 support (handle Android breaking change)

Probably easiest way to explain this is just to refer to the discussion/PR for the same change merged into `react-native-maps` - https://github.com/airbnb/react-native-maps/pull/1481.

- 9edcddf66e838f3d0ade945deba54e81937fde22 - Updated README installation instructions for Android.

Prior to this commit, there were issues with multiple versions of Firebase being included due to a bug in the Google Services plugin. This commit provides updated installation instructions that work-around the bug in the Google Services plugin.

--





If you made it all the way through that then your time was _hugely_ appreciated. You deserve a gold star! ⭐️ 


I'm certainly open to feedback and happy to correct code style issues and what-not.

Benjamin-Dobell commented:
I should probably note that this series of commits adds support for:

- `tag`

- `sound`

- `icon` (small)

- `lightsColor`, `lightsOnMs`, `lightsOffMs` (notification LED control)

- `color`
  ![color](https://cloud.githubusercontent.com/assets/482276/26714088/29c8b394-47b3-11e7-8b77-88b4ddfb6879.png)

- `largeIcon` (This image was specified as a remote URL):

  ![largeicon](https://cloud.githubusercontent.com/assets/482276/26714093/2f3d3868-47b3-11e7-9d12-08abe7ffcbaa.png)

AlimovSV commented:
Looks like `NotificationsAndroid.invalidateToken` is not available. Also moment when I call `NativeModules.WixRNNotifications.invalidateToken()` directly, the handler set by `setRegistrationTokenUpdateListener` called twice - with empty token and with newly generated token. is it ok?

AlimovSV commented:
My application uses a `react-native-device-info`. And when I migrate to your branch, i've got some conflict and:
1) Have to use `multiDexEnabled true` or app:transformClassesWithDexForDebug fail with 
```
Dex: Error converting bytecode to dex:
Cause: com.android.dex.DexIndexOverflowException: Cannot merge new index 65819 into a non-jumbo instruction!
```
2) Have to use next gradle configuration:
```
    compile(project(':react-native-device-info')) {
        exclude group: 'com.google.android.gms', module: 'play-services-gcm'
    }
    compile(project(':react-native-notifications')) {
        exclude group: 'com.google.firebase', module: 'firebase-messaging'
    }
...
    compile 'com.google.android.gms:play-services-gcm:10.2.6'
    compile 'com.google.firebase:firebase-messaging:10.2.6'
```
or on receiving notification app crashes with 
```
java.lang.NoSuchMethodError: No static method zzz ...
```
I'not familiar with java world so not sure what the best solution is.

Benjamin-Dobell commented:
@AlimovSV 

> Looks like `NotificationsAndroid.invalidateToken` is not available

Thanks for pointing that out. I must have missed it when splitting things up into feature commits. I've added an additional commit exposing the method.

> the handler set by `setRegistrationTokenUpdateListener` called twice - with empty token and with newly generated token. is it ok?

This is intentional.

When you invalidate a token with Firebase the token temporarily becomes `null` - to signify you don't have a token available. However, Firebase then sees you don't have a token and automatically generates you a new one - which triggers a second call to the token listener (with the new token).

Although unlikely, if you had connectivity issues it is possible that receiving the new token could take some time - so it's worth reporting the `null` token first to let any listeners know the old token is in fact invalid.

> 1. Have to use multiDexEnabled true or app:transformClassesWithDexForDebug fail with

This has _kind of_ has nothing to do with `react-native-notifications` or `react-native-device-info`.  I'm using both these libraries myself and haven't run into this (yet).

The issue itself is simply caused by the fact your application is very large (lots of Java methods). That's mostly due to React Native being large, however linking in other libraries results in more methods and eventually they can't all fit in one "Dex file" (a Android JVM construct). So you need to enable multidex. This is simply a limitation of the Android platform.

More details:
https://developer.android.com/studio/build/multidex.html

> 2. Have to use next gradle configuration: ...

Are you saying you needed to use both GCM and FCM, or else the app crashed?

That sounds peculiarly, if anything I would expect the two libraries to conflict. If you're using FCM you really should remove GCM.

That crash isn't very descriptive unfortunately. Are you able to recreate the crash with a debug build (or even a release with proguard disabled)? If so please provide the stack trace and a description of exactly what state the app was in and what actions were required to trigger the crash.

jjdp commented:
this is very much needed especially fcm

Benjamin-Dobell commented:
I've just fixed (or rather added) proper background notification support. Android needs a background queue, for exactly the same reasons iOS needs one. Therefore, I've added `NotificationsAndroid.consumeBackgroundQueue()` which is a *breaking change*. Events won't be delivered at all unless this is called - this is necessary functionality.

I've updated the documentation accordingly. In doing so I've also fixed addressed some extremely misleading advice / sample code. In particular the (iOS) documentation shows adding and removing event listeners inside components. This is incorrect, "cold start" background notifications will _not_ be received by your listener if they're added as part of a Component as the component hierarchy does not exist at all until your app is foregrounded.

Benjamin-Dobell commented on android/src/main/java/com/wix/reactnativenotifications/core/JsIOHelper.java:
```diff
@@ -1,38 +1,89 @@
+        if (Looper.getMainLooper() == Looper.myLooper()) {
+            emitEvent(event);
+        } else {
+            mHandler.post(new Runnable() {
```
If anybody is wondering why this is necessary, it is because our events are coming from `IntentService` (not regular `Service`) sub-classes. `IntentService` run on a dedicated background thread rather than the main thread. This becomes a problem when the app is launched in response to push notification being received. In particular, no `ReactInstanceManager`exists yet, and attempting to retrieve/create it leads to a crash inside React Native. React Native expects `ReactInstanceManager` to only be _created_ on the main thread (although it can be retrieved and utilised from any thread), internally attempting to create an Android `Handler` on an "unprepared" `Looper` leads to a runtime exception.

Benjamin-Dobell commented:
Rebased on master due to `README.md` conflicts. Initial comment has been updated with the most recent commit references.

Otherwise, I'm still awaiting feedback/review/merge 😄

d4vidi commented:
For the love of god, this PR is huge. I'm not sure when I'd be able to go over it but in any case I seriously appreciate it nonetheless...

weekian commented:
Hi! I have just created a brand new react native application and followed the steps to install react-native-notifications based on your pull request but similar to @AlimovSV, I encountered a similar error as well

`Dex: Error converting bytecode to dex:WithDexForDebug
Cause: com.android.dex.DexException: Multiple dex files define Lcom/google/android/gms/common/api/zza;
    UNEXPECTED TOP-LEVEL EXCEPTION:
com.android.dex.DexException: Multiple dex files define Lcom/google/android/gms/common/api/zza;`

In my settings.gradle,
`rootProject.name = 'testApp'
include ':react-native-notifications'
project(':react-native-notifications').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-notifications/android')
include ':app'`

In Android root-level build.gradle (based on the firebase link)
`dependencies {
        classpath 'com.android.tools.build:gradle:2.2.3'
        classpath 'com.google.gms:google-services:3.1.0'
        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }`

In app/build.gradle,
`dependencies {
    compile project(':react-native-notifications')
    compile fileTree(dir: "libs", include: ["*.jar"])
    compile "com.android.support:appcompat-v7:23.0.1"
    compile "com.facebook.react:react-native:+"  // From node_modules
}`
And at the end of the file, I added
`apply plugin: 'com.google.gms.google-services'`

It would be very helpful to include more details in setting up for receiving notifications, in my opinion. 

Please help!

Benjamin-Dobell commented:
Okay, I've investigated the multi-dex issue. It turns out it's because Google's `com.google.gms.google-services` plugin is poorly implemented.

It tries to infer the version of Firebase or Google Play Services by looking at your applications dependencies. But it doesn't check nested dependencies, only root dependencies. You guys don't have a direct dependency on Firebase or Google Play Services (I do, I've got more than just RNN), so the plugin falls back to version 9.0.0 and _injects_ Firebase 9.0.0 (an older version) into your app - so you end up with _two_ versions of Firebase - hence why the app is suddenly so large.

There's basically two ways I can think of to address this:

1. Fork `com.google.gms.google-services`.
2. In the README for `react-native-notifications` indicate that you must include Firebase messaging as an app dependency (i.e. don't attempt to automatically do so as a dependency or react-native-notifications).

`1.` sounds like a mess, and honestly I'm not even sure whether the license permits forking. `2.` doesn't sound particularly user friendly, but it'll at least work around the problem.

I'll submit a patch/commit now updating the README and removing the Firebase dependency from `react-native-notifications`.

Benjamin-Dobell commented:
@AlimovSV and @weekian, you should be good to go now. Please let me know if you run into any other issues.

weekian commented:
Hi, thank you for the swift fixes!

For your latest commit, as you removed the line compile 'com.google.firebase:firebase-messaging10.2.6', it throws up the following errors:

`node_modules\react-native-notifications\android\src\main\java\com\wix\reactnativenotifications\core\notifications\NotificationProps.java:10: error: package com.google.firebase.messaging
 does not exist
import com.google.firebase.messaging.RemoteMessage;`

and other 'cannot find symbol' errors as a result of the package not being found

Benjamin-Dobell commented:
Just a heads up, I've reported the Google Services plugin bug to Google:

https://issuetracker.google.com/issues/65311853

Once they address the bug I should be able to fix the recently reported issues in a satisfactory manner. I'll refactor and rebase once Google address the issue. It's pretty trivial in nature so hopefully they're able to make a minor version release reasonably quickly.

RangerMauve commented:
What's the main reason that this hasn't been merged in yet? It seems like a no-brainer for anybody wanting to use this library with a modern Android setup.

Benjamin-Dobell commented:
@RangerMauve Well it's actually slightly broken at the moment. The Firebase plugin has a bug which makes it impossible for libraries to use Firebase in a normal fashion. There are work around, I've been doing so in my own project, but it's a little bit clunky in terms of project setup (no code changes required). If Google don't fix it soon I'll just update the READMR explaining the integration.

jjdp commented:
i think the maintainers haven't reviewed this yet

RangerMauve commented:
@Benjamin-Dobell Will setting up Firebase the usual way before installing the plugin be sufficient?

Benjamin-Dobell commented:
@RangerMauve Yes, setup Firebase according to the official Firebase documentation:
https://firebase.google.com/docs/android/setup

However, you do _not_ need to add `firebase-core` (i.e. Google Analytics) if you don't want it.

Then the _vital_ step is to setup your Android app's dependencies as follows:

```gradle
dependencies {
    // ...

    compile(project(':react-native-notifications')) {
        exclude group: 'com.google.firebase', module: 'firebase-messaging'
    }

    com.google.firebase:firebase-messaging:10.2.1
}
```

You're welcome to use a different version than `10.2.1` if you want.

What's important here is that the app itself has a dependency on `firebase-messaging`. This is so the Google Services plugin detects that we're using Firebase Messaging and therefore _doesn't_ inject a dependency on `firebase-core:9.0.0`.

However, in order to compile `react-native-notifications` also has a dependency on `firebase-messaging`, but we're excluding it so we don't end up with duplicate dependencies which may have different versions.

RangerMauve commented:
@Benjamin-Dobell So far it seems pretty clear. We're probably going to be working on integrating with this on Monday, so I'll make sure to update you when we have it up and running and if we have any issues with the setup.

DannyvanderJagt commented:
@lidanh Any news on the progress? It seems like this PR is holding back quite some important fixes like #99. Also while searching through the list of PRs I found a lot of open PRs which didn't receive any feedback or don't have a timeline. 

At this point, combining the bugs and the lack of contributor activity, I don't find this project suitable for production anymore. Would love to help, could you let us know how much priority and resources this project has for you/your team, just so we know what we can expect? Maybe we can find some new contributors or do a sprint to get this project back into shape :)

srameshr commented:
@Benjamin-Dobell Does it cover iOS with FCM token support.

peterlazar1993 commented:
react-native-notifications has the potential to be the best notifications library on RN. Would love to see this PR merged

tommeier commented on README.md:
```diff
@@ -75,69 +75,67 @@ And the following methods to support registration and receiving notifications:
+    }
+
+    com.google.firebase:firebase-messaging:10.2.1
+	compile project(':react-native-notifications')
```
`compile 'com.google.firebase:firebase-messaging:10.2.1'`

tommeier commented on README.md:
````diff
@@ -75,69 +75,67 @@ And the following methods to support registration and receiving notifications:
+apply plugin: 'com.google.gms.google-services'
 ```
 
+**IMPORTANT**: Do NOT add `com.google.firebase:firebase-messaging` as an explicit dependency unless you wish to try override our version (see "Installation").
````
Not sure I understand this? In Installation it indicates to add as a dependency but then you're saying _not_ to here?

Benjamin-Dobell replied:
@tommeier Nyah, crud. I edited the README several times. Originally I didn't want you to have to manually add a dependency. But the bug in Google's plugin meant it was impossible. I need to fix the README.

You *do need* a dependency.

Benjamin-Dobell commented:
@tommeier Force pushed an update to the README.

tommeier commented on android/build.gradle:
```diff
@@ -20,7 +20,7 @@ android {
 dependencies {
     // Google's GCM.
-    compile 'com.google.android.gms:play-services-gcm:10.0.1'
+    compile "com.google.firebase:firebase-messaging:10.2.6"
```
which classpath are you using to match this with 10.2.6?

`classpath "com.google.gms:google-services:3.0.0"` ? its crashing when using 3.1.0 (unsure why yet)

tommeier replied:
Ah saw your note. Mine was crashing unless I included as follows: 

```
    compile 'com.google.android.gms:play-services-base:11.0.4'
    compile(project(':react-native-notifications')) {
        exclude group: 'com.google.firebase', module: 'firebase-messaging'
    }
    compile 'com.google.firebase:firebase-messaging:11.0.4'
```

tommeier commented on README.md:
```diff
@@ -274,34 +286,64 @@ NotificationsAndroid.setNotificationOpenedListener((notification) => {
-import {NotificationsAndroid, PendingNotifications} from 'react-native-notifications';
+import {NotificationsAndroid} from 'react-native-notifications';
 
 PendingNotifications.getInitialNotification()
```
`NotificationsAndroid. getInitialNotification`

tommeier commented:
FYI: We've been using this in production for a couple of weeks and is working flawlessly 🎉 Thanks @Benjamin-Dobell

CapitanRedBeard commented:
@lidanh @d4vidi Can we have an assigned reviewer for this PR to get some traction on it. I don't want to just point to a forked branch forever.

kiprijonas commented:
Is this going to be merged or is the project is dead?

jankowskip commented:
Any news about merge? It's 2018 already :boom:

rohanraarora commented:
@Benjamin-Dobell I am using React-Native-Navigation which implements Native Navigation and to integrate that with this library, I was referring to this guide (https://github.com/wix/react-native-notifications/wiki/Android:-working-with-RNN) But as there's is no `PushNotification` class in your wha would be the best way to integrate with React Native Navigation and thanks a lot for the efforts and FCM integration. 

@d4vidi it would be really great if this PR gets merged.

Benjamin-Dobell commented:
@rohanraarora I haven't gotten around to integrating Native Navigation in my RN app yet. However, I have tried to ensure my changes to `react-native-notifications` remain compatible.

The new `INotificationsApplication` method to override is:

```
ILocalNotification getLocalNotification(Context context, NotificationProps notificationProps, AppLifecycleFacade facade, AppLaunchHelper defaultAppLaunchHelper);
```

You'll want to inherit from `LocalNotification` instead of `PushNotification`.

Please let me know how you go.

batusai513 commented:
@Benjamin-Dobell Hi, great work with this one, i just have a small question, does this PR addresses: - Not firing the corresponding events when the application is in background on IOS and Android?
- Not receiving any notifications when the app is killed on Android?

Thanks in advanced.

tommeier commented:
@batusai513 I'm using this fork in production and it is working on cold, warm and hot boots from notifications without any issues.

batusai513 commented:
@tommeier Thanks for your answer, I'm trying to install this branch but i'm failing to make it work, I can't receive the deviceToken, it would be to much if i ask for your help?

here I share my config:
```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.strivehub"
    android:versionCode="1"
    android:versionName="1.0">

    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.WAKE_LOCK" />
    <uses-permission android:name="android.permission.VIBRATE" android:maxSdkVersion="18" />

    <uses-sdk
        android:minSdkVersion="16"
        android:targetSdkVersion="22" />

    <application
      android:name=".MainApplication"
      android:allowBackup="true"
      android:label="@string/app_name"
      android:icon="@mipmap/ic_launcher"
      android:theme="@style/AppTheme">
      <activity
        android:name=".SplashActivity"
        android:label="@string/app_name"
        android:configChanges="keyboard|keyboardHidden|orientation|screenSize"
        android:windowSoftInputMode="adjustResize"
        android:exported="true"
        android:theme="@style/SplashTheme">
        <intent-filter>
          <action android:name="android.intent.action.MAIN" />
          <category android:name="android.intent.category.LAUNCHER" />
        </intent-filter>
      </activity>

      <activity
        android:hardwareAccelerated="true"
        android:name=".MainActivity"
        android:exported="true"
          android:configChanges="keyboard|keyboardHidden|orientation|screenSize"
          />
      <activity android:name="com.facebook.react.devsupport.DevSettingsActivity" />
      <service
          android:name="com.wix.reactnativenotifications.fcm.FcmMessageHandlerService">
          <intent-filter>
              <action android:name="com.google.firebase.MESSAGING_EVENT"/>
          </intent-filter>
      </service>
      <service
          android:name="com.wix.reactnativenotifications.fcm.FcmInstanceIdListenerService">
          <intent-filter>
              <action android:name="com.google.firebase.INSTANCE_ID_EVENT"/>
          </intent-filter>
      </service>
      <service
        android:name="com.wix.reactnativenotifications.fcm.FcmTokenService"
        android:exported="false" />
      <meta-data
        android:name="io.fabric.ApiKey"
        android:value="4d3bb3351dd7dea22c7b176d0747a4073f8375d5" />
      <meta-data
        android:name="com.wix.reactnativenotifications.gcmSenderId"
        android:value="725801312638\0"/>
      <meta-data android:name="com.bugsnag.android.API_KEY"
             android:value="24fee5bb88f502d999112082e6d1f4b9"/>
    </application>

</manifest>
```
```gradle
// Top-level build file where you can add configuration options common to all sub-projects/modules.

buildscript {
    repositories {
        jcenter()
        maven { url 'https://maven.fabric.io/public' }

    }
    dependencies {
        classpath 'com.android.tools.build:gradle:2.2.3'
        classpath 'io.fabric.tools:gradle:1.+'
        classpath 'com.google.gms:google-services:3.1.0'

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        mavenLocal()
        jcenter()
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url "$rootDir/../node_modules/react-native/android"
        }
        maven { url 'https://maven.google.com' }
    }
}
```
```graddle
apply plugin: "com.android.application"
apply plugin: 'io.fabric'

repositories {
    maven { url 'https://maven.fabric.io/public' }
}
import com.android.build.OutputFile

/**
 * The react.gradle file registers a task for each build variant (e.g. bundleDebugJsAndAssets
 * and bundleReleaseJsAndAssets).
 * These basically call `react-native bundle` with the correct arguments during the Android build
 * cycle. By default, bundleDebugJsAndAssets is skipped, as in debug/dev mode we prefer to load the
 * bundle directly from the development server. Below you can see all the possible configurations
 * and their defaults. If you decide to add a configuration block, make sure to add it before the
 * `apply from: "../../node_modules/react-native/react.gradle"` line.
 *
 * project.ext.react = [
 *   // the name of the generated asset file containing your JS bundle
 *   bundleAssetName: "index.android.bundle",
 *
 *   // the entry file for bundle generation
 *   entryFile: "index.android.js",
 *
 *   // whether to bundle JS and assets in debug mode
 *   bundleInDebug: false,
 *
 *   // whether to bundle JS and assets in release mode
 *   bundleInRelease: true,
 *
 *   // whether to bundle JS and assets in another build variant (if configured).
 *   // See http://tools.android.com/tech-docs/new-build-system/user-guide#TOC-Build-Variants
 *   // The configuration property can be in the following formats
 *   //         'bundleIn${productFlavor}${buildType}'
 *   //         'bundleIn${buildType}'
 *   // bundleInFreeDebug: true,
 *   // bundleInPaidRelease: true,
 *   // bundleInBeta: true,
 *
 *   // whether to disable dev mode in custom build variants (by default only disabled in release)
 *   // for example: to disable dev mode in the staging build type (if configured)
 *   devDisabledInStaging: true,
 *   // The configuration property can be in the following formats
 *   //         'devDisabledIn${productFlavor}${buildType}'
 *   //         'devDisabledIn${buildType}'
 *
 *   // the root of your project, i.e. where "package.json" lives
 *   root: "../../",
 *
 *   // where to put the JS bundle asset in debug mode
 *   jsBundleDirDebug: "$buildDir/intermediates/assets/debug",
 *
 *   // where to put the JS bundle asset in release mode
 *   jsBundleDirRelease: "$buildDir/intermediates/assets/release",
 *
 *   // where to put drawable resources / React Native assets, e.g. the ones you use via
 *   // require('./image.png')), in debug mode
 *   resourcesDirDebug: "$buildDir/intermediates/res/merged/debug",
 *
 *   // where to put drawable resources / React Native assets, e.g. the ones you use via
 *   // require('./image.png')), in release mode
 *   resourcesDirRelease: "$buildDir/intermediates/res/merged/release",
 *
 *   // by default the gradle tasks are skipped if none of the JS files or assets change; this means
 *   // that we don't look at files in android/ or ios/ to determine whether the tasks are up to
 *   // date; if you have any other folders that you want to ignore for performance reasons (gradle
 *   // indexes the entire tree), add them here. Alternatively, if you have JS files in android/
 *   // for example, you might want to remove it from here.
 *   inputExcludes: ["android/**", "ios/**"],
 *
 *   // override which node gets called and with what additional arguments
 *   nodeExecutableAndArgs: ["node"],
 *
 *   // supply additional arguments to the packager
 *   extraPackagerArgs: []
 * ]
 */

project.ext.react = [
    entryFile: "index.js"
]

apply from: "../../node_modules/react-native/react.gradle"

/**
 * Set this to true to create two separate APKs instead of one:
 *   - An APK that only works on ARM devices
 *   - An APK that only works on x86 devices
 * The advantage is the size of the APK is reduced by about 4MB.
 * Upload all the APKs to the Play Store and people will download
 * the correct one based on the CPU architecture of their device.
 */
def enableSeparateBuildPerCPUArchitecture = false

/**
 * Run Proguard to shrink the Java bytecode in release builds.
 */
def enableProguardInReleaseBuilds = false

project.ext.vectoricons = [
    iconFontNames: [ "FontAwesome.ttf" ]
]

apply from: "../../node_modules/react-native-vector-icons/fonts.gradle"

android {
    compileSdkVersion 23
    buildToolsVersion '23.0.1'

    defaultConfig {
        applicationId "com.strivehub"
        minSdkVersion 16
        targetSdkVersion 22
        versionCode 3
        versionName "1.1"
        ndk {
            abiFilters "armeabi-v7a", "x86"
        }
    }
    signingConfigs {
        release {
            if (project.hasProperty('MYAPP_RELEASE_STORE_FILE')) {
                storeFile file(MYAPP_RELEASE_STORE_FILE)
                storePassword MYAPP_RELEASE_STORE_PASSWORD
                keyAlias MYAPP_RELEASE_KEY_ALIAS
                keyPassword MYAPP_RELEASE_KEY_PASSWORD
            }
        }
    }
    splits {
        abi {
            reset()
            enable enableSeparateBuildPerCPUArchitecture
            universalApk false  // If true, also generate a universal APK
            include "armeabi-v7a", "x86"
        }
    }
    buildTypes {
        debug {
            manifestPlaceholders = [excludeSystemAlertWindowPermission: "false"]
        }
        release {
            manifestPlaceholders = [excludeSystemAlertWindowPermission: "true"]
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
            signingConfig signingConfigs.release
        }
    }
    // applicationVariants are e.g. debug, release
    applicationVariants.all { variant ->
        variant.outputs.each { output ->
            // For each separate APK per architecture, set a unique version code as described here:
            // http://tools.android.com/tech-docs/new-build-system/user-guide/apk-splits
            def versionCodes = ["armeabi-v7a":1, "x86":2]
            def abi = output.getFilter(OutputFile.ABI)
            if (abi != null) {  // null for the universal-debug, universal-release variants
                output.versionCodeOverride =
                        versionCodes.get(abi) * 1048576 + defaultConfig.versionCode
            }
        }
    }
}

dependencies {
    compile project(':bugsnag-react-native')
    compile project(':react-native-keychain')
    compile project(':react-native-video')
    compile project(':react-native-orientation')
    compile project(':react-native-vector-icons')
    compile(project(':react-native-notifications')) {
        exclude group: 'com.google.firebase', module: 'firebase-messaging'
    }
    compile 'com.google.android.gms:play-services-base:11.0.4'
    compile 'com.google.firebase:firebase-messaging:11.0.4'
    compile fileTree(dir: "libs", include: ["*.jar"])
    compile "com.android.support:appcompat-v7:23.0.1"
    compile "com.facebook.react:react-native:+"  // From node_modules
    compile('com.crashlytics.sdk.android:crashlytics:2.6.8@aar') {
        transitive = true;
    }
}

// Run this once to be able to run the application with BUCK
// puts all compile dependencies into folder libs for BUCK to use
task copyDownloadableDepsToLibs(type: Copy) {
    from configurations.compile
    into 'libs'
}

buildscript {
  repositories {
    maven { url 'https://maven.fabric.io/public' }
  }
  dependencies {
    classpath 'io.fabric.tools:gradle:1.22.1'
  }
}

apply plugin: 'com.google.gms.google-services'
```

Thanks in advanced 👍

rohanraarora commented:
@Benjamin-Dobell Thanks for the info, I tried that but I am having a weird issue implementing this branch with React Native Navigation. The issue is that my app would stuck on the startup if it's being launched first time, minimising it and resuming again would launch the app properly. I also tried creating a new project with just React Native Navigation and this branch but I got the same weird behaviour. Then I replaced this branch with the master and that weird behaviour is gone but I am not able to push data-only notifications. It would show a blank notification if there's no title or body in the push message. Do you have any idea what might be issue here? I am having a hard time integrating this library with React Native Navigation. Thanks in advance.

@d4vidi Can you shed some light on the React Native Navigation part or why that's behaviour might be happening? Thanks.

tommeier commented:
@batusai513 Sorry - been away - from a quick scan, I had a similar thing where it needs to have `compile 'com.google.android.gms:play-services-base:11.+'` to work correctly. See if that helps?

jankowskip commented:
Did anyone try to customize local android notifications layout? <a href="https://github.com/wix/react-native-notifications/wiki/Android:-Layout-Customization"> Wiki</a> is depreacted after all those changes

kyryloz commented:
Hi, guys. Is this going to be merged? Would be great.

alexismoreau commented:
Hey guys, I'm using this branch with react-native-navigation, I get local notification working but the deviceToken is always null.
Hope someone can help me. here is my code : 

MainApplication.java
```
package com.timtalent.android;

import android.support.annotation.Nullable;
import android.content.Context;

import com.facebook.react.ReactPackage;
import com.reactnativenavigation.NavigationApplication;

import com.oblador.vectoricons.VectorIconsPackage;
import com.oblador.keychain.KeychainPackage;
import com.wix.reactnativenotifications.RNNotificationsPackage;
import com.wix.reactnativenotifications.core.JsIOHelper;
import com.wix.reactnativenotifications.core.AppLifecycleFacade;
import com.wix.reactnativenotifications.core.AppLaunchHelper;
import com.wix.reactnativenotifications.core.notifications.INotificationsApplication;
import com.wix.reactnativenotifications.core.notifications.ILocalNotification;
import com.wix.reactnativenotifications.core.notifications.NotificationProps;

import java.util.Arrays;
import java.util.List;

public class MainApplication extends NavigationApplication implements INotificationsApplication {

  @Override
  public boolean isDebug() {
    return BuildConfig.DEBUG;
  }

  private NotificationsLifecycleFacade notificationsLifecycleFacade;

  @Override
  public void onCreate() {
    super.onCreate();

    // Create an object of the custom facade impl
    notificationsLifecycleFacade = new NotificationsLifecycleFacade();
    // Attach it to react-native-navigation
    setActivityCallbacks(notificationsLifecycleFacade);
  }

    @Override
  public ILocalNotification getLocalNotification(Context context, NotificationProps notificationProps, AppLifecycleFacade facade, AppLaunchHelper defaultAppLaunchHelper) {
    return new AppPushNotification(
            context,
            notificationProps,
            notificationsLifecycleFacade, // Instead of defaultFacade!!!
            defaultAppLaunchHelper,
            new JsIOHelper(context)
    );
  }

  protected List<ReactPackage> getPackages() {
    // Add additional packages you require here
    // No need to add RnnPackage and MainReactPackage
    return Arrays.<ReactPackage>asList(
            new VectorIconsPackage(),
            new KeychainPackage(),
            new RNNotificationsPackage(MainApplication.this)
    );
  }

  @Nullable
  @Override
  public List<ReactPackage> createAdditionalReactPackages() {
    return getPackages();
  }

  @Override
    public String getJSMainModuleName() {
      return "index";
    }
}
```
MainActivity.java
```
package com.timtalent.android;

import com.reactnativenavigation.controllers.SplashActivity;

public class MainActivity extends SplashActivity {

}
```

AppPushNotification.java
```
package com.timtalent.android;

import android.content.Context;

import com.wix.reactnativenotifications.core.JsIOHelper;
import com.wix.reactnativenotifications.core.AppLaunchHelper;
import com.wix.reactnativenotifications.core.notifications.LocalNotification;
import com.wix.reactnativenotifications.core.notifications.NotificationProps;

public class AppPushNotification extends LocalNotification {

    public AppPushNotification(Context context, NotificationProps notificationProps, NotificationsLifecycleFacade appLifecycleFacade, AppLaunchHelper appLaunchHelper, JsIOHelper jsIoHelper) {
        super(context, notificationProps, appLifecycleFacade, appLaunchHelper);
    }
}
```
NotificationsLifecycleFacade.java

```
package com.timtalent.android;

import android.app.Activity;
import android.util.Log;

import com.facebook.react.bridge.ReactContext;
import com.reactnativenavigation.NavigationApplication;
import com.reactnativenavigation.controllers.ActivityCallbacks;
import com.reactnativenavigation.react.ReactGateway;
import com.wix.reactnativenotifications.core.AppLifecycleFacade;

import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

public class NotificationsLifecycleFacade extends ActivityCallbacks implements AppLifecycleFacade {

    private static final String TAG = NotificationsLifecycleFacade.class.getSimpleName();

    private Activity mVisibleActivity;
    private Set<AppVisibilityListener> mListeners = new CopyOnWriteArraySet<>();

    @Override
    public void onActivityResumed(Activity activity) {
        switchToVisible(activity);
    }

    @Override
    public void onActivityPaused(Activity activity) {
        switchToInvisible(activity);
    }

    @Override
    public void onActivityStopped(Activity activity) {
        switchToInvisible(activity);
    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        switchToInvisible(activity);
    }

    @Override
    public boolean isReactInitialized() {
        return NavigationApplication.instance.isReactContextInitialized();
    }

    @Override
    public ReactContext getRunningReactContext() {
        final ReactGateway reactGateway = NavigationApplication.instance.getReactGateway();
        if (reactGateway == null) {
            return null;
        }

        return reactGateway.getReactContext();
    }

    @Override
    public boolean isAppVisible() {
        return mVisibleActivity != null;
    }

    @Override
    public synchronized void addVisibilityListener(AppVisibilityListener listener) {
        mListeners.add(listener);
    }

    @Override
    public synchronized void removeVisibilityListener(AppVisibilityListener listener) {
        mListeners.remove(listener);
    }

    private synchronized void switchToVisible(Activity activity) {
        if (mVisibleActivity == null) {
            mVisibleActivity = activity;
            Log.d(TAG, "Activity is now visible ("+activity+")");
            for (AppVisibilityListener listener : mListeners) {
                listener.onAppVisible();
            }
        }
    }

    private synchronized void switchToInvisible(Activity activity) {
        if (mVisibleActivity == activity) {
            mVisibleActivity = null;
            Log.d(TAG, "Activity is now NOT visible ("+activity+")");
            for (AppVisibilityListener listener : mListeners) {
                listener.onAppNotVisible();
            }
        }
    }
}
```

```
NotificationsAndroid.localNotification({
        title: 'Local notification',
        body: 'This notification was generated by the app!',
        extra: 'data',
      });
```
is working fine but

```
NotificationsAndroid.setRegistrationTokenUpdateListener((deviceToken) => {
    console.warn(deviceToken); // NULL
  });
```

I can see in Android logcat : 
Firebase has a new token: FirebaseInstanceId:'something', token=null

Do I need to override RemoteNotification as well ?

hkung77 commented:
Is this PR still compatible with GCM? I tried the PR but was running into errors. Calling refreshToken() crashes because Firebase wasn't initialized. And the registering the token was never invoked. Can we still use the PR without google-services.json?

Can an example for the AndroidManifest be provided?

Benjamin-Dobell commented:
@hkung77 This PR is not GCM compatible as GCM is deprecated. You need to use FCM instead.

An example project:

https://github.com/Benjamin-Dobell/ReactNativeNotificationsExample

hkung77 commented:
Thanks @Benjamin-Dobell for the response! I was able to get it working setting up firebase. I too had issues getting the device token (the listener was never being called/ registered event was never emitted)... I took a look into the code and realized sJsIsReady is always false until I call consumeBackgroundQueue().

I did notice you did this in your example repository. But perhaps this should be included in the actual README for setting up token registration.

Benjamin-Dobell commented:
> I did notice you did this in your example repository. But perhaps this should be included in the actual README for setting up token registration.

It is covered:

https://github.com/Benjamin-Dobell/react-native-notifications/tree/android-overhaul#background-queue-important

Although perhaps I need to rearrange the README a bit. I'm open to suggestions 😄

hkung77 commented:
Yea Sorry about that! I saw that section after I posted the comment. Hence I tried to update it... but you beat me to it haha

VSchlattinger commented:
@Benjamin-Dobell thanks a lot for this awesome PR! It solved a lot of problems I had with using this together with FCM!

I had one problem though:

When the app is in the background (not completely closed) and I get a notification (that is not data-only), the notification automatically goes to the system tray (just like you describe in the documentation). However, when I tap on it, my RN App is not relaunched completely, but just woken up. (In your documentation you write: "When a user taps on an system tray generated notification the Android OS will actually re-launch your app's primary/launcher activity. If you're using stock React Native this effectively restarts your app."). Because it is not re-launched, but just woken up, the code, where I would usually call `NotificationsAndroid.getInitialNotification()` is not run. (The listener for NotificationOpened isn't called either, because the App was in the background.)

Maybe this happens because I'm using `react-native-navigation`?

Anyway, I solved this by adding `AppState.addEventListener("change", …)` to my code and calling `NotificationsAndroid.getInitialNotification().then(…)` whenever the state changes from "inactive" or "background" to "active". That solved this problem.

Maybe you could add something about this to the documentation just in case others will run into the same problem as me.

PS: I really hope this can be merged soon! Until then, I'm using your fork instead.

Benjamin-Dobell commented:
Just a heads up to everyone, I'm probably going to properly fork `react-native-notifications` and publish to NPM so that we're not all pointing to arbitrary git commits.

I'm also _thinking_ about separating the local notification stuff out into a separate package, so that other projects (e.g. https://github.com/invertase/react-native-firebase) can take advantage if they desire.

However, if anyone is interested, I've another branch going in my repo (https://github.com/Benjamin-Dobell/react-native-notifications/commits/firebase) that has a lot of additional functionality beyond this PR. In particular there's _proper_ Android channel support (i.e. you can create/configure channels from JS), support for [expandable notification styles](https://developer.android.com/training/notify-user/expanded), Android O fixes and WIP support for [Android actions](https://developer.android.com/training/notify-user/build-notification#Actions). It's a WIP so I don't recommend pointing your `package.json` there (as I might blast away your references), but feel free to fork and point your `package.json` to your own fork.

quinton-j commented:
@Benjamin-Dobell are you still considering creating your own npm package?  If you are, it would be awesome if you could include my PRs #88 and #90 or equivalent feature set.  It would be amazing to use npm again.

Benjamin-Dobell commented:
> @Benjamin-Dobell are you still considering creating your own npm package?

Absolutely.

Although I must admit, it's not a huge priority for me at the moment as I've been kept busy with some unrelated work. If someone else beats me to it, then I'll happily jump on board as a maintainer though.

kanso-michael commented:
@Benjamin-Dobell Thanks for your work on this, helped a great deal with some of our issues. We are currently working around a few small remaining issues and just wondered if you knew of potential causes. 
I doubt it's your fork, but your knowledge of the Android side of things would be most helpful. I'll also mention we're using React Native Navigation.

We are sending push notifications to our app to make a member aware they have received a new message. 
These push notifications are being sent as `data` only notifications.

Current problems:

(1)
When sending a notification and the app is dead, the push notification is actually starting up the app. Do you know why this from happening? 
Or is this a side effect of using `data` only push notifications?
I looked for usages of `setFullScreenIntent`, which seems to be what is occurring, but couldn't find any. 
https://developer.android.com/reference/android/app/Notification.Builder#setFullScreenIntent(android.app.PendingIntent,%20boolean)

(2)
When the app is running in the background we successfully create a `LocalNotification` in the notification centre. 
However, on clicking through to be taken into the app we see a white screen on top of the current stack. If you press the back button this screen is successfully popped off the stack and the app continues with what it needs to do.
This is likely to be a RNN issue, but just thought i'd ask if you knew of anything obvious that I might be missing.

Any help would be tremendously appreciated!

Benjamin-Dobell commented:
@kanso-michael That does sound like a bit of a doozy. Off the top of my head I'm not much use unfortunately.

> (1)
> When sending a notification and the app is dead, the push notification is actually starting up the app. Do you know why this from happening?

Your application should be started up to process the remote notification (i.e. so you can present a local notification etc.), this is the advantage of sending a data notification. However, although your app starts it _should_ be doing so in the background.

If it's not there may indeed be something slightly odd going with React Native Navigation, I'm not using React Native Navigation at present so unfortunately I'm not much help in that regard.

> I looked for usages of setFullScreenIntent, which seems to be what is occurring, but couldn't find any.
https://developer.android.com/reference/android/app/Notification.Builder#setFullScreenIntent(android.app.PendingIntent,%20boolean)

This isn't anything `react-native-notifications` is doing. Actually `setFullScreenIntent` isn't supported by `react-native-notifications`.

The only suggestion I can make, at least for testing purposes, is to try a lower priority when creating your local notification. If your notification isn't high priority then fullscreenIntents will have no impact. So if lowering the priority fixes the issue, then you're onto something. If it's happening and your notifications aren't high priority, then this is _probably_ not the issue - I suspect the latter.

> (2)
When the app is running in the background we successfully create a LocalNotification in the notification centre.
However, on clicking through to be taken into the app we see a white screen on top of the current stack.

I'm inclined to think these issues are related, if not one and the same. It sounds like when your app is starting it's pushing an activity - that might be something React Native Navigation is doing.

Unfortunately without debugging your application I probably can't be of much more help. I do actually offer consulting services, although little contracts for fixing a bug or two generally aren't particularly economical. Nonetheless, if you're _totally_ blocked by this it's an option.

kanso-michael commented:
Thanks for taking the time to read my wall of ramble. 
I seem to have found the issue in-between posting and you replying!
I have narrowed it down to being an issue with react native navigation and seem to have alleviated both described issues by adding the following code:

https://github.com/calcazar/react-native-push-notification-CE/issues/18#issuecomment-413373588

This seems to have prevented the app from automatically opening and correctly resuming from clicking on a push notification.

The last remaining issue we have is that we can't seem to get the event listeners to be correctly hit after receiving a message from a dead start.
I'm assuming it is probably something to do with RNN once again but just wanted to check that my understanding is sound. 

```
    @Override
    public void onMessageReceived(RemoteMessage message) {
        Log.d(LOGTAG, "New message from firebase");
        final NotificationProps notificationProps = NotificationProps.fromRemoteMessage(this, message);
        new RemoteNotification(this, notificationProps).onReceived();
    }
```

We are seeing this being hit in `adb logcat` which in turn should hit..

```
    private void sendReceivedEvent() {
        mJsIOHelper.sendEventToJS(NOTIFICATION_RECEIVED_EVENT_NAME, mNotificationProps.asBundle());
    }
```

Which then results in the event being fired:
`com.wix.reactnativenotifications.notificationReceived`

Now, I am of the understanding that if we implement:

```
  NotificationsAndroid.setNotificationReceivedListener((notification) => {
    console.log('NOTIFICATION_RECEIVED_LISTENER', notification)
  })
```

We should be seeing that event handler hit.
We would then schedule a new LocalNotification and the user can then press it to load up the app:

```
  NotificationsAndroid.getInitialNotification()
    .then((notification) => {
      console.log('INITIAL_NOTIFICATION', notification)
      if (notification) {
        component.onNotificationOpened(notification)
      }
    })
    .catch((err) => {
      console.error(err)
    })
```

And we can use that notification data.

If my understanding is correct as to that being what we need to do to receive the notifications from a cold start, then it is a navigation stack setup/order/timing issue that is causing this last problem. 

Thanks for reading my wall of text and greatly appreciate your previous help.

Benjamin-Dobell commented:
@kanso-michael It's mentioned a little earlier in this thread, but my recommendations for handling remote (push) notifications (such that they work from dead state) and creating local notifications is available as an example project with the most relevant code being:

https://github.com/Benjamin-Dobell/ReactNativeNotificationsExample/blob/master/js/root.js

The basic principles are also covered in the README of my fork, but seeing it in practice ought to give the most clarity.

Generally speaking I don't think you actually ever _need_ to use `NotificationsAndroid.getInitialNotification()`. Seems as I've added a proper background event queue (which you need to consume on start-up with `NotificationsAndroid.consumeBackgroundQueue()`), any listener you set with `NotificationsAndroid.setNotificationOpenedListener(onNotificationOpened)` ought to fire just fine, in which case you can handle launching from notifications there.

kanso-michael commented:
> @kanso-michael It's mentioned a little earlier in this thread, but my recommendations for handling remote (push) notifications (such that they work from dead state) and creating local notifications is available as an example project with the most relevant code being:
> 
> [Benjamin-Dobell/ReactNativeNotificationsExample:js/root.js@`master`](https://github.com/Benjamin-Dobell/ReactNativeNotificationsExample/blob/master/js/root.js)
> 
> The basic principles are also covered in the README of my fork, but seeing it in practice ought to give the most clarity.
> 
> Generally speaking I don't think you actually ever _need_ to use `NotificationsAndroid.getInitialNotification()`. Seems as I've added a proper background event queue (which you need to consume on start-up with `NotificationsAndroid.consumeBackgroundQueue()`), any listener you set with `NotificationsAndroid.setNotificationOpenedListener(onNotificationOpened)` ought to fire just fine, in which case you can handle launching from notifications there.

@Benjamin-Dobell Thank you so much for assisting, that code example is awesome. I have successfully got everything working in Android and iOS using your fork. 
Hope the maintainers get this merged at some point, fantastic work and cheers for explaining the bits I missed. 👍

ywongweb commented:
@kanso-michael 
It's been 4 months since you got this fork working on your app. How is it so far and have you encounter any other problems?
